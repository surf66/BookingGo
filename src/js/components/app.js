import React from 'react';
import Form from './form';

export default class App extends React.Component {
  render() {
    return(
      <div className="container">
        <Form />
      </div>
    );
  }
}