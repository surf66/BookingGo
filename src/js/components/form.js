import React from 'react';
import SearchField from './search-field';
import SuggestionList from './suggestion-list';

const Form = () => {
  return(
    <section>
      <form className="pickup-form">
        <h1>Let's find your ideal car</h1>
        <SearchField />
        <SuggestionList />
      </form>
    </section>
  );
}

export default Form;