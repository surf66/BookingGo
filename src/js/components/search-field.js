import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { debounce } from 'throttle-debounce';
import { SearchTermActions } from '../actions/index';

function mapDispatchToProps(dispatch) {
  return ({
    setSearchTerm: bindActionCreators(SearchTermActions.setSearchTerm, dispatch)
  })
}

export class SearchField extends React.Component {

  constructor() {
    super();

    this.handleChange = this.handleChange.bind(this);
    this.storeSearchTerm = debounce(1000, this.storeSearchTerm);
  }

  render() {
    return(
      <div className="field-container">
        <label htmlFor="pickup-location">Pick-up Location</label>
        <input type="text" id="pickup-location" name="pickup-location" placeholder="city, airport, station, region, district..." onChange={this.handleChange} />
      </div>
    );
  }

  handleChange(event) {
    let searchTerm = event.target.value;
    this.storeSearchTerm(searchTerm);
  }

  storeSearchTerm(searchTerm) {
    this.props.setSearchTerm(searchTerm);
  }
}

export default connect(null, mapDispatchToProps)(SearchField);