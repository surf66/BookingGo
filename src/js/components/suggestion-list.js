import React from 'react';
import { connect } from 'react-redux';
import SuggestionsService from '../services/suggestions-service';

function mapStateToProps(state) {
  return {
    searchTerm: state.searchTerm,
  }
}

export class SuggestionList extends React.Component {

  constructor() {
    super();

    this.state = {
      suggestions: []
    }

    this.getSuggestions = this.getSuggestions.bind(this);
  }
  
  componentWillReceiveProps(newProps) {
    this.getSuggestions(newProps.searchTerm)
  }

  render() {
    return(
      <React.Fragment>
        {this.state.suggestions.map((suggestion, index) => {
          return <p key={index}>{suggestion.name}, {suggestion.city}, {suggestion.country}</p>;
        })}
      </React.Fragment>
    );
  }

  async getSuggestions(searchTerm) {
    let suggestions = await SuggestionsService.getSuggestions(searchTerm);
    this.setState({ suggestions });
  }
}

export default connect(mapStateToProps, null)(SuggestionList);