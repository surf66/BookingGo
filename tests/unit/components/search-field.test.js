import React from 'react';
import { SearchField } from '../../../src/js/components/search-field';

test('should match snapshot', async () => {
  const component = mount(<SearchField />);

  expect(component).toMatchSnapshot();
});

test('should call setSearchTerm when handleChange is called with a searchTerm with length greater than 1', () => {
  SearchField.prototype.setSearchTerm = jasmine.createSpy();
  const component = mount(<SearchField />);
  const mockEvent = {
    target: {
      value: 'testsearchterm'
    }
  }

  component.instance().handleChange(mockEvent);

  setTimeout(() => {
    expect(SearchField.prototype.setSearchTerm).toHaveBeenCalledWith(mockEvent.target.value);
  }, 200);
});

test('should not call setSearchTerm when handleChange is called with a searchTerm with length less than 1', () => {
  SearchField.prototype.setSearchTerm = jasmine.createSpy();
  const component = mount(<SearchField />);
  const mockEvent = {
    target: {
      value: 't'
    }
  }

  component.instance().handleChange(mockEvent);

  expect(SearchField.prototype.setSearchTerm).not.toHaveBeenCalled();
});

test('should call setSearchTerm prop with searchTerm when storeSearchTerm is called', () => {
  let setSearchTermMock = jasmine.createSpy();
  const component = mount(<SearchField setSearchTerm={setSearchTermMock} />);
  const searchTerm = 'testsearchterm';

  component.instance().storeSearchTerm(searchTerm);

  setTimeout(() => {
    expect(setSearchTermMock).toHaveBeenCalledWith(searchTerm);
  }, 200);
});