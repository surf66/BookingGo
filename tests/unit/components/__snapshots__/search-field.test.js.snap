// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<SearchField>
  <div
    className="field-container"
  >
    <label
      htmlFor="pickup-location"
    >
      Pick-up Location
    </label>
    <input
      autoComplete="off"
      id="pickup-location"
      name="pickup-location"
      onChange={[Function]}
      placeholder="city, airport, station, region, district..."
      type="text"
    />
  </div>
</SearchField>
`;
